Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift

```diff
@@ -0,0 +1,117 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Async Algorithms open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+public protocol BufferStorage<Element> {
+  associatedtype Element
+  @Sendable func send(element: Element) async
+  @Sendable func fail(error: Error)
+  @Sendable func finish()
+  @Sendable func next() async -> Result<Element?, Error>
+}
+
+extension AsyncSequence where Self: Sendable {
+  public func buffer(
+    limit: UInt
+  ) -> AsyncBufferSequence<Self> {
+    self.buffer(storage: SuspendingBufferStorage(limit: limit))
+  }
+
+  public func buffer(
+    policy: QueuedBufferPolicy
+  ) -> AsyncBufferSequence<Self> {
+    self.buffer(storage: QueuedBufferStorage(policy: policy))
+  }
+
+  public func buffer<Storage: BufferStorage>(
+    storage: Storage
+  ) -> AsyncBufferSequence<Self> where Self.Element == Storage.Element {
+    AsyncBufferSequence(base: self, storage: storage)
+  }
+}
+
+public struct AsyncBufferSequence<Base: AsyncSequence & Sendable>: AsyncSequence, Sendable {
+  public typealias Element = Base.Element
+  public typealias AsyncIterator = Iterator
+
+  let base: Base
+  let send: @Sendable (Element) async -> Void
```

Comment:
so using this route will end up preventing inline optimizations (which is perhaps one of the biggest performance boons swift has to offer). I'm sure that having it be part of the generic is likely much more performant

Reply:
Hi,

I guess those 3 variants could benefit from inline optimizations right?

- using concrete types:
https://github.com/twittemb/swift-async-algorithms/blob/feature/buffer-concrete-types/Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift

- using existential:
https://github.com/twittemb/swift-async-algorithms/blob/feature/buffer-existantial/Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift

- using generics:
https://github.com/twittemb/swift-async-algorithms/blob/feature/buffer-generic-over-protocol/Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift

Reply:
I think either the generics or concrete types are the right way to go for perf. We could take the generics version and leave the construction with the storage off for now until we settle everything down. Adding it as a later addition.

Reply:
Ok I’ll work on that tomorrow and publish the PR

Reply:
I was looking at that branch further - it seems to resolve some of the issues brought up. I am going to take a look at seeing if we can reduce the surface for the `BufferStorage` down such that it can be more of a simple interface.

Reply:
Overall I'm happy with the result.
The perfs are good (+20% for the unbounded algorithm compared to the Actor version).
The AsyncStorage protocol requirements are simple.

What do you think ?